### PR TITLE
Use conda-index dependency instead of obsolete built-in index

### LINKED ISSUE #4645: Replace conda_build/index.py with standalone conda-index
### Checklist

- [X] I added a descriptive title
- [X] I searched open requests and couldn't find a duplicate

### What is the idea?

conda-build generates repodata.json during its run, using get_build_index https://github.com/conda/conda-build/blob/main/conda_build/index.py#L112

This confusing function both downloads repodata.json, and generates new repodata.json for the local package cache / the packages that were just built.

It also provides the "conda-build index" command line interface, or "conda index".

Now that we have a much faster standalone conda-index descended from conda_build/index.py, conda-build should add a dependency on conda-index, and call that instead. (The downloading parts of `get_build_index` will probably have to stay)

- [ ] package conda-index
- [ ] figure out what to do about the new CLI "conda index" (provided by standalone conda-index) versus the old CLI "conda index" provided by conda-build
- [ ] depend on conda-index
- [ ] rewrite get_build_index

### Why is this needed?

This will ensure that we have one excellent repodata.json generator.

### What should happen?

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dholth**
### Description

This replaces conda-build's index with conda-index. Fix #4645 .

Standalone conda-index caches package metadata to sqlite and fetches everything necessary for repodata.json in a single query, instead of slowly opening thousands or hundreds of thousands of individual small files. It is simpler and on the order of 10 times faster in production compared to the older `conda-build index`. Unlike the older code, it no longer reads from the previous repodata.json trying to only replace metadata related to changed packages. Instead, the data only flows in one direction from packages -> metadata cache -> repodata.json.

We have tried several strategies. This code simply keeps the old implementation for users of `conda-build index` aka `conda index` while printing many warnings to use the new package. Instead of worrying about a deprecation schedule for the missing `file` flag this code anticipates completely removing the old command and API.

Most users get the benefit automatically when conda-build uses new-conda-index internally as a side effect of doing builds.

I think it is likely that very few users need to run conda-index themselves rather than relying on hosting.

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

File: conda_build/cli/main_index.py [hunk 1]
```diff
@@ -90,8 +90,8 @@ def execute(args):
 
     api.update_index(args.dir, check_md5=args.check_md5, channel_name=args.channel_name,
                      threads=args.threads, subdir=args.subdir, patch_generator=args.patch_generator,
-                     verbose=args.verbose, progress=args.progress, hotfix_source_repo=args.hotfix_source_repo,
-                     current_index_versions=args.current_index_versions_file, index_file=args.file)
```
Comment (dholth) [inline conda_build/cli/main_index.py:94]: it might be nice to print a warning if you pass file

File: conda_build/index.py [hunk 2]
```diff
@@ -107,9 +57,18 @@ def map(self, func, *iterables):
...
+    bldpkgs_dir,
+    output_folder=None,
+    clear_cache=False,
+    omit_defaults=False,
+    channel_urls=None,
+    debug=False,
+    verbose=True,
+    locking=None,
+    timeout=None,
+):
```
Comment (dholth) [inline conda_build/index.py:71]: add comment "this is confusing" may exist in other PR's

File: conda_build/index.py [hunk 3]
```diff
@@ -226,1195 +200,46 @@ def get_build_index(subdir, bldpkgs_dir, output_folder=None, clear_cache=False,
...
+    progress=False,
+    hotfix_source_repo=None,
+    subdirs=None,
+    warn=True,
+    current_index_versions=None,
+    debug=False,
+    index_file=None,
+):
+    # conda-build calls update_index on a single subdir internally, but
+    # conda-index expects to index every subdir under dir_path
```
Comment (dholth) [inline conda_build/index.py:225]: this is possible used as an api by conda_build's index cli (although I would rather you use conda-index's new CLI with added features)

File: conda_build/index.py [hunk 4]
```diff
@@ -61,31 +31,6 @@ def map(self, func, *iterables):
...
-        "lua": "lua",
-        "nodejs": "js",
-        "perl": "perl",
-        "php": "php",
-        "ruby": "ruby",
-        "m2-base": "m2",
-        "msys2-conda-epoch": "m2w64",
-    }
-    NAMESPACE_PACKAGE_NAMES = frozenset(NAMESPACES_MAP)
-    NAMESPACES = frozenset(NAMESPACES_MAP.values())
```
Comment (dholth) [inline conda_build/index.py:87]: we could keep these maybe

File: conda_build/render.py [hunk 5]
```diff
@@ -2,36 +2,43 @@
 # SPDX-License-Identifier: BSD-3-Clause
```
Comment (dholth) [inline conda_build/render.py:2]: we might be able to pull these changes into a separate PR or rebase with the reformat

File: conda_build/utils.py [hunk 6]
```diff
@@ -1404,6 +1404,8 @@ class LoggingContext:
         "conda_build.index",
         "conda_build.noarch_python",
         "urllib3.connectionpool",
+        "conda_index",
+        "conda_index.index",
```
Comment (dholth) [inline conda_build/utils.py:1408]: this affects logging for those module

File: tests/test_index.py [hunk 7]
```diff
@@ -1,1198 +0,0 @@
-# Copyright (C) 2014 Anaconda, Inc
```
Comment (dholth) [inline tests/test_index.py:1]: 10% less test time

Comment (dholth) [reply tests/test_index.py:1]: The other conda-build tests that build packages, wind up calling get_build index and indirectly test the indexing code. Standalone conda-index has its own faster version of these tests that we delete here.

**Review by jezdez — DISMISSED** [review]
I like this PR also best.

**Comment by jezdez** [comment]
@dholth Could you rebase on top of main please?

**Comment by dholth** [comment]
> I like this PR also best.

What is best

**Comment by jezdez** [comment]
> 

I mean compared to the other options to deprecate the current index subcommand, #4798 and #4841.

File: recipe/meta.yaml [hunk 8]
```diff
@@ -28,7 +28,12 @@ requirements:
   run:
     - beautifulsoup4
     - chardet
-    - conda >=4.13
+    - conda  >=4.13
+    - conda-index
+    # - conda-verify  >=3.0.2  # optional as of CB 3.12.0
+    - contextlib2   # [py<34]
+    - enum34        # [py<34]
+    - pathlib2      # [py<3]
```
Comment (travishathaway) [inline recipe/meta.yaml:36]: Are these lines really necessary? 😅 

I didn't know we had to support Python `<3.4`...

File: conda_build/cli/main_index.py [hunk 9]
```diff
@@ -85,7 +87,7 @@ def parse_args(args):
     p.add_argument(
         "-f",
         "--file",
-        help="A file that contains a new line separated list of packages to add to repodata.",
+        help="A file that contains a new line separated list of packages to add to repodata.  Deprecated, will be removed in a future version of conda build.",
```
Comment (travishathaway) [inline conda_build/cli/main_index.py:90]: I know that conda-build hasn't officially adopted the [Conda Release Schedule](https://github.com/conda-incubator/ceps/blob/main/cep-8.md) but I still think it would be a good idea to provide a version number when this will be removed. Otherwise, it probably never will 🤷.

**Review by travishathaway — COMMENTED** [review]
I like the direction that this pull request is going in, but I think we need to be more specific about when conda-index will be officially removed from conda-build.

Eventually I would like to see the this command removed from the CLI of conda-build and replaced by adding conda-index as a proper subcommand plugin (https://github.com/conda/conda-index/issues/81).

Is there a way that we can add our intention to do this either via the release notes or in the documentation? It doesn't have to be completed in this pull request, and I would be more than happy to help out.

File: conda_build/cli/main_index.py [hunk 10]
```diff
@@ -96,6 +98,11 @@ def parse_args(args):
 def execute(args):
     _, args = parse_args(args)
 
+    if args.file:
+        warnings.warn("--file is no longer implemented.")
+    if args.hotfix_source_repo:
+        warnings.warn("--hotfix-source-repo is no longer implemented.")
+
```
Comment (jaimergp) [inline conda_build/cli/main_index.py:105]: These aren't available in the new conda-index, right? Are there equivalent functionalities? Should people be warned with a `DeprecationWarning` before we do remove these?

**Review by jaimergp — COMMENTED** [review]
Thanks Daniel!

I have some questions:

- What's the story for users of the `conda index` subcommand in this PR? Are they silently being upgraded to the `conda-index` rewrite? If that's the case, what backwards incompatible changes should they expect?
- Same but for users of `conda_build.api`.

File: conda_build/index.py [hunk 11]
```diff
@@ -40,6 +39,7 @@
 #  BAD BAD BAD - conda internals
 from conda.core.subdir_data import SubdirData
 from conda.models.channel import Channel
+from conda_index.index import update_index as _update_index
```
Comment (jaimergp) [inline conda_build/index.py:42]: Users have been silently updated to the new `conda-index` via this import. Have we considered deprecating this module first with a warning? What's the story in the future? Will `conda_build.index` always exist? If not, should we add a module-level warning (bis)?

Comment (dholth) [reply conda_build/index.py:42]: How do we add a module warning without triggering it when conda-build imports

File: conda_build/cli/main_index.py (hunk 10 above)
Comment (dholth) [reply conda_build/cli/main_index.py:105]: `--hotfix` was removed before I started working on conda-build and `--file` isn't implemented in new conda-index. The replacement is that new conda-index is fast.

File: recipe/meta.yaml (hunk 8 above)
Comment (dholth) [reply recipe/meta.yaml:36]: No I didn't touch the recipe as part of this PR.

File: conda_build/cli/main_index.py (hunk 9 above)
Comment (dholth) [reply conda_build/cli/main_index.py:90]: > Thanks Daniel!
> 
> I have some questions:
> 
> * What's the story for users of the `conda index` subcommand in this PR? Are they silently being upgraded to the `conda-index` rewrite? If that's the case, what backwards incompatible changes should they expect?
> * Same but for users of `conda_build.api`.

Yes, everyone is upgraded to the new conda-index with the possible exception of people who import parts of this file that we have left in for backwards compatibility to follow a deprecation policy that this project has not officially adopted.

Users should expect equivalent indexing with a sqlite cache instead of a many-small-files cache.

In this PR we try to map all the API calls to forwarded calls to new-conda-index.

File: conda_build/index.py (hunk 11 above)
Comment (jezdez) [reply conda_build/index.py:42]: I think the goal would be eventually to replace this code completely with the implementation in conda-index via an own subcommand. There are two groups of users here:

- End-users of the index subcommand, they can continue to use it, and they shouldn't see any difference to the built-in index feature.
- Makers who import from `conda_build.index` in their scripts/tools and rely on the Python API. They'll need to be provided with enough information and a rollout plan to make the necessary changes.

I only worry about the second group, and how to reach them. They'll unlikely read the conda-build changelog, so we'll need to let them know via `DeprecationWarning` instead. I don't think it has to be a module-level warning, maybe a function wrapper like `conda.deprecations` would suffice.

**Comment by jezdez** [comment]
> I like the direction that this pull request is going in, but I think we need to be more specific about when conda-index will be officially removed from conda-build.
> 
> Eventually I would like to see the this command removed from the CLI of conda-build and replaced by adding conda-index as a proper subcommand plugin ([conda/conda-index#81](https://github.com/conda/conda-index/issues/81)).

💯 

> Is there a way that we can add our intention to do this either via the release notes or in the documentation? It doesn't have to be completed in this pull request, and I would be more than happy to help out.

I think there is a way to do this, providing a deprecation warning for importers of the `conda_build.index` module (or its functions etc) and an additional note about the removal in N numbers of releases.

I'd suggest in 2 releases (a fast version of CEP 9), e.g. in 3.27.0 by current nomenclature, which would coincide with the conda deprecation release 23.9.0.

**Comment by jezdez** [comment]
Bonus points if we could do the same for conda skeleton 😬 But maybe that's asking for too much!

**Comment by dholth** [comment]
We need to finish this PR. Somehow it has been delayed at least since January over 

> Bonus points if we could do the same for conda skeleton 😬 But maybe that's asking for too much!

@travishathaway if you want to make conda-index a cli plugin, PR's are welcome against https://github.com/conda/conda-index although I'm not sure how to resolve the redesigned click-CLI differences (we thought about giving it a different name but haven't found the right synonym for 'index' ('index2'?)

**Comment by dholth** [comment]
I've updated this PR to leave the `conda-build index` command aka `conda index` alone; it will use the old implementation. conda's deprecation handler is copied in to avoid depending on very new conda.

Only conda-build's own indexing should use the new implementation. That way normal use gets the benefit and API or CLI use gets compatible behavior in case the `--file` feature is needed.

Users without those unusual requirements should use standalone `conda-index` for fast indexing, repodata.json.zst support etc.

File: conda_build/index.py (hunk 11 above)
Comment (dholth) [reply conda_build/index.py:42]: Updated to expose all the old API's with the old implementation, while printing warnings.

File: conda_build/build.py [hunk 12]
```diff
@@ -37,7 +37,7 @@
 from conda_build.config import Config
 from conda_build.create_test import create_all_test_files
 from conda_build.exceptions import CondaBuildException, DependencyNeedsBuildingError
-from conda_build.index import get_build_index, update_index
+from conda_build.index import _delegated_update_index, get_build_index
```
Comment (dholth) [inline conda_build/build.py:40]: better name welcome

File: conda_build/cli/main_index.py [hunk 13]
```diff
@@ -96,6 +99,10 @@ def parse_args(args):
 def execute(args):
     _, args = parse_args(args)
 
+    warnings.warn(
+        "conda-build index is deprecated. Use standalone conda-index package."
```
Comment (jezdez) [inline conda_build/cli/main_index.py:103]:
```suggestion
        "conda-build index is deprecated. Use the standalone conda-index package instead."
```

File: news/4645-use-conda-index [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+### Bug fixes
+
+* <news item>
+
+### Deprecations
+
+* `conda-build index --file` is no longer supported.
+* `conda-build index --hotfix-source-repo` is removed.
+* Prefer `python -m conda_index`, instead of `conda-build index` or `conda
+  index`, to use faster indexing code.
```
Comment (jezdez) [inline news/4645-use-conda-index:14]: I think the users should prefer the use of the index subcommand long-term, when conda-index has implemented a plugin based subcommand. In the meantime let's not send them down the `-m conda_index` rabbit hole here in the changelog, and instead point them to the conda-index docs. That'll allow us to have one source of truth about how to interact with conda-index long-term, its own docs.

```suggestion
* Prefer the [standalone conda-index package](https://conda.github.io/conda-index/),
  instead of `conda-build index` or `conda index`, to use faster indexing code.
```

File: recipe/meta.yaml (hunk 8 above)
Comment (jezdez) [inline recipe/meta.yaml:36]:
```suggestion
```

File: recipe/meta.yaml [hunk 15]
```diff
@@ -28,7 +28,12 @@ requirements:
   run:
     - beautifulsoup4
     - chardet
-    - conda >=4.13
+    - conda  >=4.13
```
Comment (jezdez) [inline recipe/meta.yaml:31]:
```suggestion
    - conda >=4.13
```

File: tests/requirements.txt [hunk 16]
```diff
@@ -3,7 +3,8 @@
 anaconda-client
 beautifulsoup4
 chardet
-conda >=4.13
+conda >= 4.13
```
Comment (jezdez) [inline tests/requirements.txt:6]:
```suggestion
conda >=4.13
```
IIRC matchspec parsing can be brittle, was this change intentional?

**Review by jezdez — APPROVED** [review]
I like the idea of us setting a clear target on what comes next when we actually remove the index code from conda-build: a conda-build 4.0.0 release. It's a backward-incompatible change to remove the index code, no matter the amount of warning we do ahead of time.

**Comment by jaimergp** [comment]
> I'm not sure how to resolve the redesigned click-CLI differences

@dholth, is this about potential conflicts with `conda-build`'s `conda index` command? I guess we can go the packaging route and use a `run_constrained` rule to make sure there are now CLI collisions. AKA a subcommand-providing `conda-index` version will only be compatible with a conda-build that has dropped the `conda index` subcommand.

File: tests/requirements.txt (hunk 16 above)
Comment (dholth) [reply tests/requirements.txt:6]: No

**Comment by dholth** [comment]
> > I'm not sure how to resolve the redesigned click-CLI differences
> 
> @dholth, is this about potential conflicts with `conda-build`'s `conda index` command? I guess we can go the packaging route and use a `run_constrained` rule to make sure there are now CLI collisions. AKA a subcommand-providing `conda-index` version will only be compatible with a conda-build that has dropped the `conda index` subcommand.

This is about updating shell scripts that run conda index.

File: news/4645-use-conda-index (hunk 14 above)
Comment (dholth) [reply news/4645-use-conda-index:14]: A reminder that `conda index` is currently an alias for `conda-build index`

File: news/4645-use-conda-index [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+* Depend on standalone conda-index instead of bundled indexing code. (#4690)
+
+### Bug fixes
+
+* <news item>
+
+### Deprecations
+
+* `conda-build index --file` is no longer supported.
+* `conda-build index --hotfix-source-repo` is removed.
```
Comment (dholth) [inline news/4645-use-conda-index]: Now we anticipate removing the entire `conda index` instead of removing the two features not implemented in standalone `conda-index`

File: tests/test_api_consistency.py [hunk 18]
```diff
@@ -214,4 +218,5 @@ def test_api_update_index():
         False,
         None,
         None,
+        inspect._empty,
```
Comment (dholth) [inline tests/test_api_consistency.py:221]: It looks like we aren't too far off from the `getfullargspec` version, `parameters` appears to include the positional and `**kwargs` parameters. `getfullargspec()` doesn't return anything for a `@wraps` function...

File: conda_build/deprecations.py [hunk 19]
```diff
@@ -0,0 +1,275 @@
+# Copyright (C) 2014 Anaconda, Inc
```
Comment (travishathaway) [inline conda_build/deprecations.py:1]: Is there a reason we're not importing this directly from `conda`?

Comment (travishathaway) [reply conda_build/deprecations.py:1]: Thanks for clarifying 👌.

**Review by travishathaway — APPROVED** [review]
Seems reasonable enough. My only reservation about the prior implementation was not having a clear deprecation path (i.e. which version we remove it in), but it looks like we've addressed that with the inclusion of our handy deprecation decorator.

**Review by kenodegard — APPROVED** [review]
I like this and think it strikes a good enough balance between deprecating/providing backports and switching to the new conda-index codes

My only worry is that conda-build 4.0.0 feels too much like the fabled conda 5.0.0 which never happened but I don't think that should block this

**Review by jezdez — APPROVED** [review]
